File: core/src/test/classes/start_system_test.cpp
//This file is part of Bertini 2.0.
//
//start_system_test.cpp is free software: you can redistribute it and/or modify
//it under the terms of the GNU General Public License as published by
//the Free Software Foundation, either version 3 of the License, or
//(at your option) any later version.
//
//start_system_test.cpp is distributed in the hope that it will be useful,
//but WITHOUT ANY WARRANTY; without even the implied warranty of
//MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//GNU General Public License for more details.
//
//You should have received a copy of the GNU General Public License
//along with start_system_test.cpp.  If not, see <http://www.gnu.org/licenses/>.
//

//  start_system_test.cpp
//
//  copyright 2015
//  Daniel Brake
//  University of Notre Dame
//  ACMS
//  Spring, Summer 2015

//start_system_test.cpp
//


#include <boost/test/unit_test.hpp>



#include "start_system.hpp"

using System = bertini::System;
using Var = std::shared_ptr<bertini::Variable>;
using VariableGroup = bertini::VariableGroup;


template<typename T> using Vec = Eigen::Matrix<T, Eigen::Dynamic, 1>;
template<typename T> using Mat = Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>;

double threshold_clearance_d = 1e-15;
double threshold_clearance_mp = 1e-27;
unsigned FUNCTION_TREE_TEST_MPFR_DEFAULT_DIGITS = 30;

BOOST_AUTO_TEST_SUITE(system_class)





BOOST_AUTO_TEST_CASE(make_total_degree_system_linear)
{
	bertini::System sys;
	Var x = std::make_shared<bertini::Variable>("x"), y = std::make_shared<bertini::Variable>("y");

	VariableGroup v;
	v.push_back(x); v.push_back(y);

	sys.AddVariableGroup(v);
	sys.AddFunction(x + y - 1);
	sys.AddFunction(x - 0.5*y - 1);


	bertini::start_system::TotalDegree TD(sys);

	auto d = TD.Degrees();

	BOOST_CHECK_EQUAL(d.size(),2);
	if (d.size()==2)
	{
		BOOST_CHECK_EQUAL(d[0],1);
		BOOST_CHECK_EQUAL(d[1],1);
	}
	
	
	BOOST_CHECK_EQUAL(TD.NumVariables(),2);

	BOOST_CHECK_EQUAL(TD.NumStartPoints(), 1);
}



BOOST_AUTO_TEST_CASE(make_total_degree_system_quadratic)
{
	bertini::System sys;
	Var x = std::make_shared<bertini::Variable>("x"), y = std::make_shared<bertini::Variable>("y");

	VariableGroup v;
	v.push_back(x); v.push_back(y);

	sys.AddVariableGroup(v);
	sys.AddFunction(x*y + y - 1);
	sys.AddFunction(x*x - 0.5*y - x*y);


	bertini::start_system::TotalDegree TD(sys);

	auto d = TD.Degrees();

	BOOST_CHECK_EQUAL(d.size(),2);
	if (d.size()==2)
	{
		BOOST_CHECK_EQUAL(d[0],2);
		BOOST_CHECK_EQUAL(d[1],2);
	}
	
	
	BOOST_CHECK_EQUAL(TD.NumVariables(),2);
	BOOST_CHECK_EQUAL(TD.NumStartPoints(), 4);
}



BOOST_AUTO_TEST_CASE(linear_total_degree_start_system)
{
	bertini::System sys;
	Var x = std::make_shared<bertini::Variable>("x"), y = std::make_shared<bertini::Variable>("y");

	VariableGroup vars;
	vars.push_back(x); vars.push_back(y);

	sys.AddVariableGroup(vars);  
	sys.AddFunction(y+1);
	sys.AddFunction(x+y+bertini::Pi());

	bertini::start_system::TotalDegree TD(sys);

	auto deg = TD.Degrees();

	BOOST_CHECK_EQUAL(deg.size(),2);
	if (deg.size()==2)
	{
		BOOST_CHECK_EQUAL(deg[0],1);
		BOOST_CHECK_EQUAL(deg[1],1);
	}

	Vec<dbl> vals(2);
	vals << dbl(1.0),dbl(1.0);

	auto sysvals = TD.Eval(vals);

	for (unsigned ii = 0; ii < 2; ++ii)
		BOOST_CHECK( abs(sysvals(ii) - (1.0 - dbl(TD.RandomValue(ii)))) < threshold_clearance_d);



	auto J = TD.Jacobian(vals);

	BOOST_CHECK_EQUAL(J(0,0),1.0);
	BOOST_CHECK_EQUAL(J(0,1),0.0);

	BOOST_CHECK_EQUAL(J(1,0),0.0);
	BOOST_CHECK_EQUAL(J(1,1),1.0);


	vals << dbl(0.0),dbl(0.0);

	sysvals = TD.Eval(vals);

	for (unsigned ii = 0; ii < 2; ++ii)
		BOOST_CHECK(abs(sysvals(ii)+dbl(TD.RandomValue(ii))) < threshold_clearance_d);
	

	J = TD.Jacobian(vals);

	BOOST_CHECK_EQUAL(J(0,0),1.0);
	BOOST_CHECK_EQUAL(J(0,1),0.0);

	BOOST_CHECK_EQUAL(J(1,0),0.0);
	BOOST_CHECK_EQUAL(J(1,1),1.0);

	BOOST_CHECK_EQUAL(TD.NumStartPoints(), 1);
}







BOOST_AUTO_TEST_CASE(quadratic_cubic_quartic_total_degree_start_system)
{
	bertini::System sys;
	Var x = std::make_shared<bertini::Variable>("x"), y = std::make_shared<bertini::Variable>("y"), z = std::make_shared<bertini::Variable>("z");

	VariableGroup vars;
	vars.push_back(x); vars.push_back(y); vars.push_back(z);

	sys.AddVariableGroup(vars);  
	sys.AddFunction(y+x*y + 0.5);
	sys.AddFunction(pow(x,3)+x*y+bertini::E());
	sys.AddFunction(pow(x,2)*pow(y,2)+x*y*z*z - 1);

	bertini::start_system::TotalDegree TD(sys);

	auto deg = TD.Degrees();

	BOOST_CHECK_EQUAL(deg.size(),3);
	if (deg.size()==3)
	{
		BOOST_CHECK_EQUAL(deg[0],2);
		BOOST_CHECK_EQUAL(deg[1],3);
		BOOST_CHECK_EQUAL(deg[2],4);
	}

	Vec<dbl> vals(3);
	vals << dbl(1.0),dbl(1.0),dbl(1.0);

	auto sysvals = TD.Eval(vals);

	for (unsigned ii = 0; ii < 3; ++ii)
		BOOST_CHECK( abs(sysvals(ii) - (1.0 - dbl(TD.RandomValue(ii)))) < threshold_clearance_d);


	auto J = TD.Jacobian(vals);

	BOOST_CHECK_EQUAL(J(0,0),2.0);
	BOOST_CHECK_EQUAL(J(0,1),0.0);
	BOOST_CHECK_EQUAL(J(0,2),0.0);

	BOOST_CHECK_EQUAL(J(1,0),0.0);
	BOOST_CHECK_EQUAL(J(1,1),3.0);
	BOOST_CHECK_EQUAL(J(1,2),0.0);

	BOOST_CHECK_EQUAL(J(2,0),0.0);
	BOOST_CHECK_EQUAL(J(2,1),0.0);
	BOOST_CHECK_EQUAL(J(2,2),4.0);



	vals << dbl(0.0),dbl(0.0),dbl(0.0);

	sysvals = TD.Eval(vals);

	for (unsigned ii = 0; ii < 3; ++ii)
		BOOST_CHECK(abs(sysvals(ii)+dbl(TD.RandomValue(ii))) < threshold_clearance_d);

	J = TD.Jacobian(vals);

	BOOST_CHECK_EQUAL(J(0,0),0.0);
	BOOST_CHECK_EQUAL(J(0,1),0.0);
	BOOST_CHECK_EQUAL(J(0,2),0.0);

	BOOST_CHECK_EQUAL(J(1,0),0.0);
	BOOST_CHECK_EQUAL(J(1,1),0.0);
	BOOST_CHECK_EQUAL(J(1,2),0.0);

	BOOST_CHECK_EQUAL(J(2,0),0.0);
	BOOST_CHECK_EQUAL(J(2,1),0.0);
	BOOST_CHECK_EQUAL(J(2,2),0.0);


	BOOST_CHECK_EQUAL(TD.NumStartPoints(), 24);
}




BOOST_AUTO_TEST_CASE(quadratic_cubic_quartic_start_points)
{
	bertini::System sys;
	Var x = std::make_shared<bertini::Variable>("x"), y = std::make_shared<bertini::Variable>("y"), z = std::make_shared<bertini::Variable>("z");

	VariableGroup vars;
	vars.push_back(x); vars.push_back(y); vars.push_back(z);

	sys.AddVariableGroup(vars);  
	sys.AddFunction(y+x*y + 0.5);
	sys.AddFunction(pow(x,3)+x*y+bertini::E());
	sys.AddFunction(pow(x,2)*pow(y,2)+x*y*z*z - 1);

	bertini::start_system::TotalDegree TD(sys);

	for (size_t ii = 0; ii < TD.NumStartPoints(); ++ii)
	{
		auto start = TD.StartPoint<dbl>(ii);
		auto function_values = TD.Eval(start);

		for (size_t jj = 0; jj < function_values.size(); ++jj)
			BOOST_CHECK(abs(function_values(jj)) < threshold_clearance_d);
	}

	boost::multiprecision::mpfr_float::default_precision(FUNCTION_TREE_TEST_MPFR_DEFAULT_DIGITS);

	for (size_t ii = 0; ii < TD.NumStartPoints(); ++ii)
	{
		auto start = TD.StartPoint<mpfr>(ii);
		auto function_values = TD.Eval(start);

		for (size_t jj = 0; jj < function_values.size(); ++jj)
		{
			BOOST_CHECK(abs(function_values(jj)) < threshold_clearance_mp);
		}
	}

}



BOOST_AUTO_TEST_CASE(quadratic_cubic_quartic_all_the_way_to_final_system)
{
	bertini::System sys;
	Var x = std::make_shared<bertini::Variable>("x"), y = std::make_shared<bertini::Variable>("y"), z = std::make_shared<bertini::Variable>("z");

	VariableGroup vars;
	vars.push_back(x); vars.push_back(y); vars.push_back(z);

	sys.AddVariableGroup(vars);  
	sys.AddFunction(y+x*y + 0.5);
	sys.AddFunction(pow(x,3)+x*y+bertini::E());
	sys.AddFunction(pow(x,2)*pow(y,2)+x*y*z*z - 1);

	bertini::start_system::TotalDegree TD(sys);

	Var t = std::make_shared<bertini::Variable>("t");

	auto final_mixed_sum = (1-t) * sys + t * TD;
	final_mixed_sum.AddPathVariable(t);
	final_mixed_sum.Homogenize();

	BOOST_CHECK_EQUAL(final_mixed_sum.NumVariables(),4);
	BOOST_CHECK(final_mixed_sum.IsHomogeneous());
	BOOST_CHECK(final_mixed_sum.IsPolynomial());

}




BOOST_AUTO_TEST_SUITE_END()

File: python/test/classes/start_system_test.py
# This file is part of Bertini 2.
#
# python/test/classes/start_system_test.py is free software: you can
# redistribute it and/or modify it under the terms of the GNU General Public
# License as published by the Free Software Foundation, either version 3 of
# the License, or (at your option) any later version.
#
# This file is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# this file.  If not, see <http://www.gnu.org/licenses/>.
#
#  Copyright(C) Bertini2 Development Team
#
#  See <http://www.gnu.org/licenses/> for a copy of the license,
#  as well as COPYING.  Bertini2 is provided with permitted
#  additional terms in the b2/licenses/ directory.

# individual authors of this file include:
#   silviana amethyst

"""Tests for the TotalDegree start system.

Mirrors C++ tests in core/test/classes/start_system_test.cpp:
  - system_class/make_total_degree_system_linear
  - system_class/make_total_degree_system_quadratic
  - system_class/linear_total_degree_start_system
  - system_class/quadratic_cubic_quartic_total_degree_start_system
  - system_class/quadratic_cubic_quartic_start_points
  - system_class/quadratic_cubic_quartic_start_points_homogenized_patched
  - system_class/total_degree_start_system_precision_16
  - system_class/total_degree_start_system_homogenized_patched_precision_16
"""

import numpy as np
import pytest

import bertini
from bertini import System, VariableGroup
from bertini.function_tree.symbol import Variable
from bertini.function_tree import *
import bertini.system.start_system as ss

import bertini.multiprec as mp
from bertini.multiprec import Complex as mpfr_complex


@pytest.fixture
def linear_system():
    """Single-variable linear system: 2x - 3."""
    x = Variable("x")
    s = System()
    vg = VariableGroup(); vg.append(x)
    s.add_variable_group(vg)
    s.add_function(2 * x - 3)
    return s


@pytest.fixture
def quad_cubic_quartic_system():
    """3-variable system with degrees (2, 3, 4): 24 start points."""
    x = Variable("x"); y = Variable("y"); z = Variable("z")
    s = System()
    vg = VariableGroup(); vg.append(x); vg.append(y); vg.append(z)
    s.add_variable_group(vg)
    s.add_function(x**2 + y**2 - 1)
    s.add_function(x**3 - z)
    s.add_function(y**4 + z**2 - 2)
    return s, x, y, z


# ---------------------------------------------------------------------------
# num_start_points and degrees
# ---------------------------------------------------------------------------

def test_total_degree_linear(linear_system):
    """Linear system has exactly 1 start point.

    Mirrors start_system_test.cpp/make_total_degree_system_linear.
    """
    td = ss.TotalDegree(linear_system)
    assert td.num_start_points() == 1
    assert list(td.degrees()) == [1]


def test_total_degree_quad_cubic_quartic(quad_cubic_quartic_system):
    """Degrees (2,3,4) yield 24 start points.

    Mirrors start_system_test.cpp/make_total_degree_system_quadratic
    (generalised to the 3-variable case).
    """
    s, *_ = quad_cubic_quartic_system
    td = ss.TotalDegree(s)
    assert list(td.degrees()) == [2, 3, 4]
    assert td.num_start_points() == 24


# ---------------------------------------------------------------------------
# start_point_d — double precision start points
# ---------------------------------------------------------------------------

def test_start_points_evaluate_to_zero_double(quad_cubic_quartic_system):
    """All 24 double start points evaluate to ~0 on the TD system.

    Mirrors start_system_test.cpp/quadratic_cubic_quartic_start_points.
    """
    s, *_ = quad_cubic_quartic_system
    td = ss.TotalDegree(s)
    n = td.num_start_points()
    assert n == 24
    for i in range(n):
        p = td.start_point_d(i)
        ev = td.eval(p)
        assert np.linalg.norm(ev) < 1e-5, f"start_point_d({i}) residual {np.linalg.norm(ev)}"


def test_total_degree_jacobian_at_111(quad_cubic_quartic_system):
    """Jacobian diagonal at (1,1,1) equals the degree vector (2,3,4).

    Mirrors start_system_test.cpp/quadratic_cubic_quartic_total_degree_start_system.
    The TD Jacobian at (1,1,1) should be diag(2,3,4) since the TD functions
    are d_i * x_i^{d_i - 1}.
    """
    s, *_ = quad_cubic_quartic_system
    td = ss.TotalDegree(s)
    td.differentiate()
    pt = np.array([complex(1, 0)] * 3)
    J = td.eval_jacobian(pt)
    assert abs(J[0, 0] - complex(2, 0)) < 1e-14
    assert abs(J[1, 1] - complex(3, 0)) < 1e-14
    assert abs(J[2, 2] - complex(4, 0)) < 1e-14


# ---------------------------------------------------------------------------
# start_point_mp — multiprecision start points
# ---------------------------------------------------------------------------

def test_start_points_evaluate_to_zero_mp(quad_cubic_quartic_system):
    """All 24 mp start points evaluate to ~0 on the TD system.

    Mirrors start_system_test.cpp/quadratic_cubic_quartic_start_points (mp variant).
    """
    s, *_ = quad_cubic_quartic_system
    td = ss.TotalDegree(s)
    n = td.num_start_points()
    for i in range(n):
        p = td.start_point_mp(i)
        ev = td.eval(p)
        norm = max(mp.abs(v) for v in ev)
        assert norm < 1e-5, f"start_point_mp({i}) residual {norm}"


# ---------------------------------------------------------------------------
# Homogenized + patched total degree system
# ---------------------------------------------------------------------------

def test_total_degree_homogenized_patched(quad_cubic_quartic_system):
    """After homogenize + auto_patch the TD system is hom. and patched.

    Mirrors start_system_test.cpp/quadratic_cubic_quartic_start_points_homogenized_patched.
    """
    s, *_ = quad_cubic_quartic_system
    s.homogenize()
    s.auto_patch()
    assert s.is_homogeneous()
    assert s.is_patched()

    td = ss.TotalDegree(s)
    td.homogenize()
    assert td.is_homogeneous()
    assert td.is_patched()

    n = td.num_start_points()
    assert n == 24
    for i in range(n):
        p = td.start_point_d(i)
        ev = td.eval(p)
        assert np.linalg.norm(ev) < 1e-5, f"start_point_d({i}) residual {np.linalg.norm(ev)}"


# ---------------------------------------------------------------------------
# Precision of start points
# ---------------------------------------------------------------------------

def test_start_point_precision_16(quad_cubic_quartic_system):
    """All start points inherit the current default precision (16).

    Mirrors start_system_test.cpp/total_degree_start_system_precision_16.
    """
    bertini.default_precision(16)

    s, *_ = quad_cubic_quartic_system
    td = ss.TotalDegree(s)
    n = td.num_start_points()
    for i in range(n):
        p = td.start_point_mp(i)
        assert all(c.precision == 16 for c in p), (
            f"start_point_mp({i}) has wrong precision: {[c.precision for c in p]}"
        )
